src/Scoreboard.h

```diff
@@ -52,11 +58,32 @@ class cObjective
 		otStatEntityKilledBy
 	};
 
-	// tolua_end
+	struct eType
+	{
+		eType(eCriteria a_Criteria) : m_Criteria(a_Criteria), m_SubCriteria(0) {}
+		eType(eCriteria a_Criteria, int a_SubCriteria) : m_Criteria(a_Criteria), m_SubCriteria(a_SubCriteria) {}
 
-	static AString TypeToString(eType a_Type);
+		bool operator ==(eType a_Other)
+		{
+			return (m_Criteria == a_Other.m_Criteria) && (m_SubCriteria == a_Other.m_SubCriteria);
+		}
+
+		eCriteria m_Criteria;
+		int m_SubCriteria;
+	};
 
 	static eType StringToType(const AString & a_Name);
+	static AString TypeToString(eType a_Type);
```

Comment:
Don't export these two to Lua API directly, but rather manually convert them to the two enum values:
```lua
cObjective:StringToType("...") -- returns number, number
cObjective:TypeToString(otDummy, 0)
```

Reply:
`RegisterObjective`'s API-facing signature would have to change to take either two ints or a string (or, plugins would have to manually construct an eType (now called Criteria in my local branch)). Its current use is:

```
Scoreboard:RegisterObjective("list", "List Scores", cObjective:StringToType("dummy"))
```

Same for `cScoreboard::ForEachObjectiveWith`.

Reply:
Yes, and it would actually still work - Lua allows functions to return multiple values and when such a function is called as the last "parameter" of another function, all of the return values are passed:
```lua
local function giveValues()
  return "hello", 7, "worlds"
end

-- These two print the same thing:
print("hello", 7, "worlds")
print(giveValues())
````

Reply:
Huh, that's pretty useful (as a language feature). Today I learned.

That said, I'm not entirely sure we want to do that here. If in the future we wanted to give RegisterObjective another (optional) parameter, existing plugins would break (since it's no longer the last two parameters).

It may make sense to export cObjective::Criteria (née cObjective::eType) to the lua API (maybe even as cObjectiveCriteria). Then also instead of the static methods for converting to/from strings, those could be a Criteria method and constructor, respectively:

```
local Objective = Scoreboard:RegisterObjective("list", "List Scores", cObjectiveCriteria("dummy"))
LOG("Has criteria ", Objective:GetCriteria():ToString())
```

Reply:
I'm hesitant to introduce a new API type just for wrapping two numbers representing some type. Would it be reasonable to pack it into a single number (`enum` even better)? Since there is a list of all possible values, it would make sense to represent it as a single enum, possibly with helper functions to break it apart into components?

Reply:
Vanilla supports 1,791 different criterion for scoreboards (22 criterion classes + all the subcriteria), explicitly enumerating them would be tricky. But, it could be done as a bitfield (16 bits + 16 bits would work, and fit it into an int). Then lua sees it as a single number. Helper functions could be exposed so lua can query the class, convert to strings, and so forth.